### DISCUSSION TITLE
Dagster Databricks Step Launcher package vs. script mode

### DISCUSSION THREAD

**Discussion by geoHeil** [dagster-io/dagster#18426] [Q&A] [answered]
When developing dagster pipelines there are two ways how a code location might be structured:

- package: with pip install -e (editable mode) and  imports like `from mypackage import foo` (which gives net clear references)
- script: no package relative imports (https://realpython.com/absolute-vs-relative-python-imports/)

The issue is: It looks like the step launcher needs to have the full dagster contect (inluding packages not used in the step operation - otherwise it fails with a missing package error).

There is a hacky way how to potentially fix it to upload packages on demand:

https://github.com/dagster-io/dagster/blob/master/python_modules/libraries/dagster-databricks/dagster_databricks/_test_utils.py#L32-L43

What would be a good way to make the upload smooth? And the interoperation with the step launcher better/more direct?

I currently think to use dagster with the external pipe mode and maybe to follow a similar-ish approach

In fact I am specifying the folders like:

```

"local_pipeline_package_path": str(Path(__file__).parent.parent.parent.parent),
```

When checking the paths: `my_library.egg-info` and `my.egg-info` should both be part of the ZIP file which is generated.

```
ls ../../../../my_library
my_library          my_library.egg-info my_library_tests    setup.py
❯ ls ../../../../my
my          my.egg-info my_dbt      my_tests    pyproject.toml setup.cfg      setup.py
❯ ls ../../../../my_library
my_library          my_library.egg-info my_library_tests    setup.py
```

Still `import my_library` fails with a missing package error.

**Comment by geoHeil** [comment]
The orignal hacky approach https://github.com/dagster-io/dagster/discussions/18383#discussioncomment-7728640 is requiring 2 means of autentication (once the `WorkspaceClient` and then for the `dbfs` CLI.

Instead:

```
with open(local_file_path, 'rb') as file:
    encoded_string = base64.b64encode(file.read()).decode('utf-8')
databricks_client.dbfs.put(path=dbfs_destination_path, contents=encoded_string, overwrite=True)

```

needs only a single auth. to upload the file.

E2E this might look like:

```
from dagster_databricks import PipesDatabricksClient
from databricks.sdk import WorkspaceClient
import os
import base64

from databricks.sdk import WorkspaceClient
import os
import subprocess

databricks_client = WorkspaceClient(
            host=os.environ["DATABRICKS_HOST"],
            token=os.environ["DATABRICKS_TOKEN"],
        )

pipe_client = PipesDatabricksClient(databricks_client)

############
# handling a single script
# Path to your local file and destination path in DBFS
local_file_path = "/path/to/external_databricks_script.py"
dbfs_destination_path = "/external_databricks/external_databricks_script.py"

with open(local_file_path, 'rb') as file:
    encoded_string = base64.b64encode(file.read()).decode('utf-8')
databricks_client.dbfs.put(path=dbfs_destination_path, contents=encoded_string, overwrite=True)

databricks_client.dbutils.fs.ls('/external_databricks')
```

This meand the problem of how to handle libraries is still remaining.

Building the library is outlined in the initial (hacky) approach:

```
mylib_path = "/path/to/my_library"
target_platform = "linux_x86_64"
subprocess.check_call(["python", os.path.join(mylib_path, "setup.py"), "bdist_wheel"], cwd=mylib_path)
```

uploading would work in a similar way:

```
wheel_file_path = "/path/to/my_library/dist/my_library-0.0.0-py3-none-any.whl"
dbfs_wheel_path = "/customlibs/ascii-0.0.0-py3-none-any.whl"
with open(wheel_file_path, 'rb') as file:
    encoded_string = base64.b64encode(file.read()).decode('utf-8')
databricks_client.dbfs.put(path=dbfs_destination_path, contents=encoded_string, overwrite=True)
```

This leaves the question open how to install the whell to the cluster. https://docs.databricks.com/en/libraries/workspace-libraries.html#install-a-workspace-library-onto-a-cluster or how dagster-pipes might be able to directly attach the library when launching (Workspace libraries are deprecated. Databricks recommends adding libraries directly to compute or job tasks.)

```
cluster_id = "<your-cluster-id>"
install_lib_request = {
    "cluster_id": cluster_id,
    "libraries": [{"whl": dbfs_wheel_path}]
}
databricks_client.libraries.install(install_lib_request)

```

might be a potential draft to install a library after the launch. However, this would mix business logic of the script of dagster-pipes with infra-management. Is there a way how dagster pipes could install the library directly? or request installation on cluster launch?

**Reply by geoHeil** [reply]
GPT 4 suggests a draft answer for this:

```
from databricks import DatabricksAPIClient

# Initialize the Databricks client
databricks_url = "https://<your-databricks-workspace-url>"
personal_access_token = "<your-personal-access-token>"
client = DatabricksAPIClient(host=databricks_url, token=personal_access_token)

# Define the job with library
job_settings = {
    "name": "Example Job with Library",
    "tasks": [{
        "task_key": "example_task",
        "description": "Example task with a library",
        "libraries": [{"whl": "dbfs:/path/to/your/library.whl"}],
        "spark_python_task": {
            "python_file": "dbfs:/path/to/your/python/script.py",
        }
    }]
}

# Create the job
response = client.jobs.create(job_settings)
job_id = response.get("job_id")

print(f"Job created with ID: {job_id}")

```

I still have to explore this for real

**Reply by geoHeil** [reply]
@smackesey The suggestion above almost works: `task = jobs.SubmitTask.from_dict(` is the correct way. The custom libraries can be neatly specified:

```
"libraries": [
                {"pypi": {"package": "dagster-pipes"}},
                {"whl": "dbfs:/customlibs/my_first-0.0.0-py3-none-any.whl"},
                {"whl": "dbfs:/customlibs/my_second-0.0.0-py3-none-any.whl"},
            ],
```

This leaves 2 points open:

- how to isolate dependencies for the libraries uploaded to spark. Probably I wll create a shared library without any other dependencies and put that code there and import this for databricks and the normal dagster jobs
- how to neatly swap between the databricks and local mode https://github.com/dagster-io/dagster/discussions/18383#discussioncomment-7718779

For the swapping: Are there any examples available to orient myself on?

**Reply by geoHeil** [reply]
And https://databricks-sdk-py.readthedocs.io/en/latest/wait.html the long running mode where I was submitting the job (not only the task) did not work for me.

In fact, I could successfully submit the job - but pipes had some error. Probably https://docs.dagster.io/guides/dagster-pipes/dagster-pipes-details-and-customization some customization might be able to solve that part

**Reply by alangenfeld** [reply]
> but pipes had some error

What was the error you encountered?

**Reply by geoHeil** [reply, marked as answer]
It was complaining about receiving a job instead a task and some dict not matching.

I think what you said here https://github.com/dagster-io/dagster/discussions/18459#discussioncomment-7792530 is an answer and customization would be required.
